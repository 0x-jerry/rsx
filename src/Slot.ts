import { isFn, remove } from '@0x-jerry/utils'
import { type AnyProps, useRawChildren } from '.'
import type {
  ExposedFunctionalComponent,
  FunctionalComponent,
} from './defineComponent'
import { defineComponentName } from './helper'
import { ComponentNode } from './nodes/ComponentNode'
import type { NodeElement } from './nodes/shared'
import { TextNode } from './nodes/TextNode'

export type Slot<T extends AnyProps = AnyProps> =
  ExposedFunctionalComponent<T> & {
    __slot: true
    name: string
  }

function isSlot(o: unknown): o is Slot {
  return isFn(o) && '__slot' in o
}

/**
 * Filter and remove raw children by Component
 *
 * @private
 *
 * @param factory
 * @returns
 */
export function useSlot<T extends AnyProps>(factory: SlotFactory<T>) {
  const children = useRawChildren()

  let dynamicSlot: unknown
  const staticSlotContents: NodeElement[][] = []

  const _ = remove(children, (child) => {
    const isStaticSlot = ComponentNode.is(child) && child.tag === factory

    if (isStaticSlot && child.children) {
      staticSlotContents.push(child.children)
      return true
    }

    const isDynamicSlot =
      TextNode.is(child) &&
      isSlot(child.content) &&
      child.content.name === factory.__slot_factory.name

    if (isDynamicSlot) {
      if (!dynamicSlot) dynamicSlot = child.content
      return true
    }

    return false
  })

  const slot = dynamicSlot ?? (() => staticSlotContents)

  return slot as Slot<T>
}

interface SlotFactory<T extends AnyProps = AnyProps>
  extends FunctionalComponent {
  (props: FunctionalComponent<T>): Slot<T>
  __slot_factory: {
    name: string
  }
}

function _isSlotFactory(o: unknown): o is SlotFactory {
  return isFn(o) && '__slot_factory' in o
}

export function defineNamedSlot<T extends AnyProps = AnyProps>(
  slotName: string,
) {
  const factory = ((slot) => {
    const SlotComponent = slot as Slot

    SlotComponent.__slot = true

    defineComponentName(SlotComponent, slotName)

    return SlotComponent
  }) as SlotFactory<T>

  factory.__slot_factory = { name: slotName }

  return factory
}
