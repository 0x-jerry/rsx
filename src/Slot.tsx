import { isObject, type Optional, remove } from '@0x-jerry/utils'
import type { Ref } from '@vue/reactivity'
import {
  $,
  type AnyProps,
  defineComponent,
  provide,
  toBindingRefs,
  useContext,
  useRawChildren,
} from '.'
import type { DNodeContext } from './context'
import type { FunctionalComponent } from './defineComponent'
import { createFragment } from './nodes/AnchorNode'
import { ComponentNode, createComponentNode } from './nodes/ComponentNode'
import { createNativeNode } from './nodes/NativeNode'
import { type NodeElement, NodeType } from './nodes/shared'
import { createTextNode } from './nodes/TextNode'
import { defineComponentName } from './test'

export type Slot = FunctionalComponent

interface SlotProp {
  _slot_prop: true
  key: string
  slot: Slot
}

const SlotDataKey = '__slot_data'

const SlotImpl = defineComponent((props, children) => {
  provide(SlotDataKey, props)

  return createFragment(children)
})

/**
 * Filter and remove raw children by Component
 *
 * @private
 *
 * @param SlotComponent
 * @returns
 */
function useRawChildrenBySlot(SlotComponent: Slot) {
  const children = useRawChildren()

  const Contents = remove(
    children,
    (child) => ComponentNode.is(child) && child.tag === SlotComponent,
  )

  return Contents as ComponentNode[]
}

export function defineNamedSlot<T extends AnyProps = AnyProps>(name?: string) {
  const Component = SlotImpl.bind({})

  if (name) {
    defineComponentName(Component, name)
  }

  const SlotComponent = defineSlotProps<T>(Component)

  return SlotComponent
}

type SlotWithData<T extends AnyProps> = Slot & { [key in keyof T]: Ref<T[key]> }

function defineSlotProps<T extends AnyProps>(Slot: Slot) {
  const ProxiedSlot = new Proxy(Slot, {
    get(_target, p, _receiver) {
      const key = p as string

      if (key === 'name') {
        return _target[key]
      }

      return createSlotProp(key)
    },
  })

  return ProxiedSlot as SlotWithData<T>

  function createSlotProp(key: string): SlotProp {
    return {
      _slot_prop: true,
      key,
      slot: ProxiedSlot,
    }
  }
}

export function useSlot<T extends AnyProps>(SlotComponent: SlotWithData<T>) {
  const rawChildren = useRawChildrenBySlot(SlotComponent)

  const Slot = defineComponent<T>((props) => {
    const clonedChildren = rawChildren.map((child) => {
      const newNode = cloneSlotContentNode(child)

      Object.assign(newNode.props, toBindingRefs(props))

      return newNode
    })

    return <>{clonedChildren}</>
  })

  return Slot
}

function cloneSlotContentNode<T extends NodeElement>(node: T): T {
  switch (node.type) {
    case NodeType.Component: {
      return createComponentNode(
        node.tag,
        cloneSlotContentProps(node.props),
        cloneSlotContentNodes(node.children),
      ) as T
    }

    case NodeType.Anchor: {
      return createFragment(cloneSlotContentNodes(node.children)) as T
    }

    case NodeType.Native: {
      return createNativeNode(
        node.tag,
        cloneSlotContentProps(node.props || {}),
        cloneSlotContentNodes(node.children),
      ) as T
    }

    case NodeType.Text: {
      return createTextNode(cloneSlotProp(node.content)) as T
    }

    default:
      throw new Error(`invalid node type`)
  }
}

function cloneSlotContentNodes(nodes?: NodeElement[]) {
  return nodes?.map((node) => cloneSlotContentNode(node))
}

function cloneSlotContentProps(props: AnyProps) {
  return Object.fromEntries(
    Object.entries(props).map(([key, value]) => {
      return [key, cloneSlotProp(value)]
    }),
  )
}

function cloneSlotProp(value: unknown) {
  return isSlotRef(value) ? createSlotPropRef(value) : value
}

function isSlotRef(o: unknown): o is SlotProp {
  return isObject(o) && '_slot_prop' in o
}

function createSlotPropRef(prop: SlotProp) {
  let ctx: Optional<DNodeContext>
  return $(() => {
    if (!ctx) {
      ctx = resolveContext(prop)
    }

    const data = (ctx.ex?.[SlotDataKey] as any)?.[prop.key]
    return data
  })
}

function resolveContext(prop: SlotProp) {
  let ctx: DNodeContext | undefined | null = useContext()

  while (ctx) {
    if (ctx._node?.tag === prop.slot) {
      return ctx
    }
    ctx = ctx.parent
  }

  throw Error(`Slot data should only used inside the corelative Slot instance.`)
}
