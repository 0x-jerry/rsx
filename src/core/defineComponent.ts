import { isFn } from '@0x-jerry/utils'
import type { Merge } from 'type-fest'
import type { ToMaybeRef } from '.'
import type { DefineProps } from './props'

export interface PropOption<T = any> {
  type?: T
  default?: T | (() => T)
}

export type PropsType<T = any> = Record<string, T | PropOption<T>>

export interface ComponentOption<Props extends PropsType = {}> {
  props?: Props
  toRefs?: boolean
}

interface CommonProps {
  class?: any
  [key: string]: any
}

export type FunctionalComponent<P extends PropsType = {}> = (
  props: Merge<P, CommonProps>,
  children?: any[],
) => JSX.Element

/**
 * without option
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * with option
 * @param opt
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P>,
  impl: FunctionalComponent<DefineProps<P>>,
): FunctionalComponent<ToMaybeRef<P>>
/**
 * implement
 */
export function defineComponent<P extends PropsType>(
  opt: ComponentOption<P> | FunctionalComponent,
  impl?: FunctionalComponent,
): FunctionalComponent {
  if (isFn(opt)) return opt

  // todo check props type
  return impl!
}

export const dc = defineComponent
