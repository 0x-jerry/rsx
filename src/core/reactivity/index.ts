export { computed, reactive, ref } from '@vue/reactivity'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './utils'
export * from './watch'
