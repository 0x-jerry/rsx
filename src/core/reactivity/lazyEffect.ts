import type { Fn } from '@0x-jerry/utils'
import {
  ReactiveEffect,
  type ReactiveEffectOptions,
  type ReactiveEffectRunner,
} from '@vue/reactivity'

export function lazyEffect(fn: Fn, options?: ReactiveEffectOptions) {
  if ((fn as ReactiveEffectRunner).effect instanceof ReactiveEffect) {
    fn = (fn as ReactiveEffectRunner).effect.fn
  }

  const e = new ReactiveEffect(fn)
  if (options) {
    Object.assign(e, options)
  }

  const runner = e.run.bind(e) as ReactiveEffectRunner
  runner.effect = e
  return runner
}
