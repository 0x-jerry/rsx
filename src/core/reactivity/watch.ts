import { isArray, isFn, type Promisable } from '@0x-jerry/utils'
import {
  effect,
  isRef,
  type ReactiveEffectOptions,
  type Ref,
  stop,
  unref,
} from '@vue/reactivity'
import { queueJob } from '../scheduler'

export type StopWatcher = () => void

export type TriggerFn<T> = (newVal: T, oldValue?: T) => Promisable<void>

export interface WatchOption extends Omit<ReactiveEffectOptions, 'lazy'> {
  immediate?: boolean
}

/**
 * used to check oldValue
 */
const initOldValue = Symbol()

export function watch<T>(
  getter: Ref<T> | (() => T),
  fn: TriggerFn<T>,
  option?: WatchOption,
): StopWatcher {
  let oldVal: any = initOldValue
  let triggered = false

  const getterIsRef = isRef(getter)

  const runner = effect(effectFn, {
    ...option,
    lazy: false,
    scheduler: () => {
      queueJob(runner)
    },
  })

  return () => stop(runner)

  function effectFn() {
    let newVal: any

    if (getterIsRef) {
      newVal = getter.value
    } else if (isFn(getter)) {
      newVal = getter()
    }

    newVal = unref(newVal)

    if (isLooseEq(oldVal, newVal)) {
      return
    }

    if (option?.immediate || (!option?.immediate && triggered)) {
      fn(newVal, oldVal === initOldValue ? undefined : oldVal)
      oldVal = newVal
    }

    triggered = true
  }
}

function isLooseEq(a: unknown, b: unknown) {
  if (a instanceof Date && b instanceof Date) {
    return a.getTime() === b.getTime()
  }

  if (isArray(a) && isArray(b) && a.length === b.length) {
    return a.every((v, idx) => v === b[idx])
  }

  return a === b
}
