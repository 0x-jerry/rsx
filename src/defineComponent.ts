import type { EmptyObject } from '@0x-jerry/utils'
import type { DefineProps } from './props'
import type { IsStartWithCapitalizedLetter, MaybeRef } from './types'

export type PropsType<T = any> = Record<string, T>

type ToPropRefs<T extends {}> = {
  /**
   * Start with capitalized letter means this prop is not used as reactive prop
   */
  [key in keyof T]: key extends IsStartWithCapitalizedLetter<key>
    ? T[key]
    : MaybeRef<T[key]>
}

/**
 * @private
 */
export type FunctionalComponent<P extends PropsType = any> = (
  props: P,
  children?: any[],
) => unknown

type _NoInferProps<P extends EmptyObject> = DefineProps<P> & Record<string, any>

export type ExposedFunctionalComponent<P extends PropsType = any> =
  FunctionalComponent<ToPropRefs<P>>

/**
 * Auto calculate props from generic type P
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: NoInfer<FunctionalComponent<_NoInferProps<P>>>,
): ExposedFunctionalComponent<P>

/**
 * Infer props from impl parameter
 * @param impl
 */
export function defineComponent<P extends PropsType>(
  impl: FunctionalComponent<P>,
): ExposedFunctionalComponent<P>

/**
 * implement
 */
export function defineComponent(impl?: any): any {
  // todo check props type
  return impl
}

export const dc = defineComponent
