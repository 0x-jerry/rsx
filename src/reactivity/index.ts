export { computed, reactive, ref } from '@vue/reactivity'
export * from './bindingRef'
export * from './effect'
export { nextTick } from './scheduler'
export * from './types'
export * from './watch'
