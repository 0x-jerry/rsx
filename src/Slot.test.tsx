import { defineComponent } from './defineComponent'
import { $, nextTick, ref } from './reactivity'
import { defineNamedSlot, useSlot } from './Slot'
import { mountTestApp } from './test'

describe('Slot', () => {
  it('get slot content', () => {
    const AImpl = defineComponent((_, children) => {
      const Title = useSlot(A.Title)
      const Content = useSlot(A.Content)

      return (
        <div class="A">
          <div class="title">
            <Title />
          </div>
          <span>-</span>
          <div class="content">
            <Content />
          </div>
          {children}
        </div>
      )
    })

    const A = Object.assign(AImpl, {
      Title: defineNamedSlot('A.Title'),
      Content: defineNamedSlot('A.Title'),
    })

    const App = () => {
      return (
        <div>
          <A>
            <span>1</span>
            <A.Title>title</A.Title>
            <span>2</span>
            <A.Content>content</A.Content>
            <span>3</span>
          </A>
        </div>
      )
    }

    const el = mountTestApp(App)

    expect(el).toMatchSnapshot('html')
  })

  it('pass data through slot', async () => {
    const count = ref(0)

    const AImpl = defineComponent((_, children) => {
      const Title = useSlot(A.Title)

      return (
        <div class="A">
          <div class="title">
            <Title count={count} />
          </div>
          {children}
        </div>
      )
    })

    const A = Object.assign(AImpl, {
      Title: defineNamedSlot<{ count: number }>('A.Title'),
    })

    const App = () => {
      return (
        <div>
          <A>
            <span>1</span>
            <A.Title>title {A.Title.count}</A.Title>
          </A>
        </div>
      )
    }

    const rootEl = mountTestApp(App)

    let text = rootEl.querySelector('.title')?.textContent
    expect(text).toBe('title 0')

    count.value = 1
    await nextTick()

    text = rootEl.querySelector('.title')?.textContent
    expect(text).toBe('title 1')
  })

  it('multiple instance', async () => {
    const AImpl = defineComponent<{ count: number }>((props, children) => {
      const Title = useSlot(A.Title)

      return (
        <div class="A">
          <div class="title">
            <Title count={$(() => props.count)} />
          </div>
          {children}
        </div>
      )
    })

    const A = Object.assign(AImpl, {
      Title: defineNamedSlot<{ count: number }>('A.Title'),
    })

    const App = () => {
      return (
        <div>
          <A count={1}>
            <span>1</span>
            <A.Title>title {A.Title.count}</A.Title>
          </A>
          <A count={2}>
            <span>1</span>
            <A.Title>title {A.Title.count}</A.Title>
          </A>
        </div>
      )
    }

    const rootEl = mountTestApp(App)

    const text = rootEl
      .querySelectorAll('.title')
      .values()
      .map((item) => item.textContent)
      .toArray()
    expect(text).eql(['title 1', 'title 2'])
  })

  it('sub child of a component', async () => {
    const AImpl = defineComponent<{ count: number }>((props, children) => {
      const Title = useSlot(A.Title)

      return (
        <div class="A">
          <div class="title">
            <Title count={$(() => props.count)} />
          </div>
          {children}
        </div>
      )
    })

    const A = Object.assign(AImpl, {
      Title: defineNamedSlot<{ count: number }>('A.Title'),
    })

    const B = defineComponent((_, children) => {
      return <div>{children}</div>
    })

    const App = () => {
      return (
        <div>
          <A count={1}>
            <span>1</span>
            <A.Title>
              <B>title {A.Title.count}</B>
            </A.Title>
          </A>
          <A count={2}>
            <span>1</span>
            <A.Title>title {A.Title.count}</A.Title>
          </A>
        </div>
      )
    }

    const rootEl = mountTestApp(App)

    const text = rootEl
      .querySelectorAll('.title')
      .values()
      .map((item) => item.textContent)
      .toArray()
    expect(text).eql(['title 1', 'title 2'])
  })
})
