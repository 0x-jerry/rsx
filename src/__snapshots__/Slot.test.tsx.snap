// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Slot > get slot content > html 1`] = `
<div
  id="app"
>
  <div>
    <div
      class="A"
    >
      <div
        class="title"
      >
        title
        <!--Fragment-->
        <!--Fragment-->
        <!--Fragment-->
      </div>
      <span>
        -
      </span>
      <div
        class="content"
      >
        content
        <!--Fragment-->
        <!--Fragment-->
        <!--Fragment-->
      </div>
      <span>
        1
      </span>
      <span>
        2
      </span>
      <span>
        3
      </span>
      <!--Fragment-->
    </div>
  </div>
</div>
`;
